// @peersyst/near-mobile-signer/dist/src/data-access/api/service/schemas/$SignerRequestDto.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.$SignerRequestDto = void 0;
exports.$SignerRequestDto = {
    properties: {
        id: {
            type: 'string',
            isRequired: true,
        },
        status: {
            type: 'Enum',
            isRequired: true,
        },
        network: {
            type: 'Enum',
            isRequired: true,
        },
        signerAccountId: {
            type: 'string',
            isRequired: true,
        },
        requests: {
            type: 'array',
            contains: {
                properties: {},
            },
            isRequired: true,
        },
        dAppMetadata: {
            type: 'DAppMetadataDto',
        },
        txHash: {
            type: 'array',
            contains: {
                type: 'string',
            },
        },
        createdAt: {
            type: 'string',
            isRequired: true,
            format: 'date-time',
        },
        updatedAt: {
            type: 'string',
            isRequired: true,
            format: 'date-time',
        },
    },
};
