// @peersyst/near-mobile-signer/dist/src/data-access/api/service/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.SignerRequestApi = exports.$SignRequest = exports.$SignMessageRequestPayload = exports.$SignMessageRequestDto = exports.$SignerTransactionDto = exports.$SignerRequestStatusDto = exports.$SignerRequestDto = exports.$DAppMetadatRequest = exports.$DAppMetadataDto = exports.$CreateSignMessageRequest = exports.$CreateSignerRequest = exports.OpenAPI = exports.CancelError = exports.CancelablePromise = exports.ApiError = void 0;
var ApiError_1 = require("./core/ApiError");
Object.defineProperty(exports, "ApiError", {
    enumerable: true,
    get: function() {
        return ApiError_1.ApiError;
    }
});
var CancelablePromise_1 = require("./core/CancelablePromise");
Object.defineProperty(exports, "CancelablePromise", {
    enumerable: true,
    get: function() {
        return CancelablePromise_1.CancelablePromise;
    }
});
Object.defineProperty(exports, "CancelError", {
    enumerable: true,
    get: function() {
        return CancelablePromise_1.CancelError;
    }
});
var OpenAPI_1 = require("./core/OpenAPI");
Object.defineProperty(exports, "OpenAPI", {
    enumerable: true,
    get: function() {
        return OpenAPI_1.OpenAPI;
    }
});
var _CreateSignerRequest_1 = require("./schemas/$CreateSignerRequest");
Object.defineProperty(exports, "$CreateSignerRequest", {
    enumerable: true,
    get: function() {
        return _CreateSignerRequest_1.$CreateSignerRequest;
    }
});
var _CreateSignMessageRequest_1 = require("./schemas/$CreateSignMessageRequest");
Object.defineProperty(exports, "$CreateSignMessageRequest", {
    enumerable: true,
    get: function() {
        return _CreateSignMessageRequest_1.$CreateSignMessageRequest;
    }
});
var _DAppMetadataDto_1 = require("./schemas/$DAppMetadataDto");
Object.defineProperty(exports, "$DAppMetadataDto", {
    enumerable: true,
    get: function() {
        return _DAppMetadataDto_1.$DAppMetadataDto;
    }
});
var _DAppMetadatRequest_1 = require("./schemas/$DAppMetadatRequest");
Object.defineProperty(exports, "$DAppMetadatRequest", {
    enumerable: true,
    get: function() {
        return _DAppMetadatRequest_1.$DAppMetadatRequest;
    }
});
var _SignerRequestDto_1 = require("./schemas/$SignerRequestDto");
Object.defineProperty(exports, "$SignerRequestDto", {
    enumerable: true,
    get: function() {
        return _SignerRequestDto_1.$SignerRequestDto;
    }
});
var _SignerRequestStatusDto_1 = require("./schemas/$SignerRequestStatusDto");
Object.defineProperty(exports, "$SignerRequestStatusDto", {
    enumerable: true,
    get: function() {
        return _SignerRequestStatusDto_1.$SignerRequestStatusDto;
    }
});
var _SignerTransactionDto_1 = require("./schemas/$SignerTransactionDto");
Object.defineProperty(exports, "$SignerTransactionDto", {
    enumerable: true,
    get: function() {
        return _SignerTransactionDto_1.$SignerTransactionDto;
    }
});
var _SignMessageRequestDto_1 = require("./schemas/$SignMessageRequestDto");
Object.defineProperty(exports, "$SignMessageRequestDto", {
    enumerable: true,
    get: function() {
        return _SignMessageRequestDto_1.$SignMessageRequestDto;
    }
});
var _SignMessageRequestPayload_1 = require("./schemas/$SignMessageRequestPayload");
Object.defineProperty(exports, "$SignMessageRequestPayload", {
    enumerable: true,
    get: function() {
        return _SignMessageRequestPayload_1.$SignMessageRequestPayload;
    }
});
var _SignRequest_1 = require("./schemas/$SignRequest");
Object.defineProperty(exports, "$SignRequest", {
    enumerable: true,
    get: function() {
        return _SignRequest_1.$SignRequest;
    }
});
var SignerRequestApi_1 = require("./services/SignerRequestApi");
Object.defineProperty(exports, "SignerRequestApi", {
    enumerable: true,
    get: function() {
        return SignerRequestApi_1.SignerRequestApi;
    }
});
